signac/core/validators.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Validators for SyncedCollection API.
+
+Validators should be callable. They should act recursively for nested data and
+they should not return any values, only raise errors.
+"""
+
+from collections.abc import Mapping
+from collections.abc import Sequence
+
+from ..errors import InvalidKeyError
+from ..errors import KeyTypeError
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+
+def no_dot_in_key(data):
```

Comment:
Actually I intended for the validators should be exposed as part of the API. That would allow users to freely apply validators in the event that we start prepackaging more of them.

Reply:
@vyasr Making them public sets a much higher bar for us to ensure their correctness, performance, and API stability. I suggest we make them internal for now, and reconsider making them public in the future. It's very hard to retract or alter public APIs, as you know.

Reply:
While I agree with you in principle, I think in this case we are actually better serving this interest by making these functions public. I think we should be designing the synced collections as a complete abstraction layer relative to the rest of signac, which means that signac should only consume the public API of these classes. Enforcing that requirement will force us to critically evaluate what should be public and private in the synced collections. Since `SyncedCollection.add_validator` is public, it makes sense for us to provide examples of what validators should look like; if we foresaw that changing, that would indicate that `add_validator` would also have to be internal, and that doesn't make any sense to me since (as I mentioned in [this comment](https://github.com/glotzerlab/signac/pull/378#discussion_r481559972)) signac will be applying the `JSON_format_validator` to all backends.  Exposing `add_validator` means the fundamental validator API must be clearly established before we can release this. Regarding correctness and performance, I would argue that issues with correctness constitute bugs that we'll have to fix anyway, and we don't make any performance guarantees with respect to validation, and optimizations are internal changes.

Since we've decided not to deal with removing non-str support for keys as of yet, we will be able to merge synced collections and make a 1.x release, so we'll have time to evaluate the API and if necessary make changes for 2.0. Alternatively, if you are particularly wary, we could leave them internal for now and make them public at the point of a 2.0 release. However, I would need a very strong reason to keep them internal after a 2.0 release; the need to do that would indicate in my view a significant flaw the design.